libcudacxx/include/cuda/std/__cccl/assert.h

```diff
@@ -0,0 +1,123 @@
+#  ifndef CCCL_ENABLE_ASSERTIONS
+#    define CCCL_ENABLE_ASSERTIONS
+#  endif // !CCCL_ENABLE_ASSERTIONS
+#endif // _CCCL_ENABLE_DEBUG_MODE
+
+//! Ensure that we switch on host assertions when all assertions are enabled
+#ifndef CCCL_ENABLE_HOST_ASSERTIONS
+#  ifdef CCCL_ENABLE_ASSERTIONS
+#    define CCCL_ENABLE_HOST_ASSERTIONS
+#  endif // CCCL_ENABLE_ASSERTIONS
+#endif // !CCCL_ENABLE_HOST_ASSERTIONS
+
+//! Ensure that we switch on device assertions when all assertions are enabled
+#ifndef CCCL_ENABLE_DEVICE_ASSERTIONS
+#  ifdef CCCL_ENABLE_ASSERTIONS
+#    define CCCL_ENABLE_DEVICE_ASSERTIONS
+#  endif // CCCL_ENABLE_ASSERTIONS
+#endif // !CCCL_ENABLE_DEVICE_ASSERTIONS
+
+//! Use internal nvcc implementation on device or the host library for clang-cuda
+#ifdef __CUDA_ARCH__
+#  ifdef CCCL_ENABLE_DEVICE_ASSERTIONS
+#    if defined(_CCCL_CUDA_COMPILER_NVCC) //! Use __assert_fail to implement device side asserts
+#      define _CCCL_ASSERT_IMPL(expression, message)       \
+        __builtin_expect(static_cast<bool>(expression), 1) \
+          ? (void) 0                                       \
+          : __assert_fail(message, __FILE__, __LINE__, __func__)
+#    else // ^^^ _CCCL_CUDA_COMPILER_NVCC ^^^ / vvv !_CCCL_CUDA_COMPILER_NVCC vvv
+#      if __has_include(<__assert>) // libc++ uses _LIBCPP_ASSERT from <__assert>
+#        include <__assert>
+#        define _CCCL_ASSERT_IMPL(expression, message) _LIBCPP_ASSERT(expression, message)
+#      else // libstdc++ uses __glibcxx_assert from <bits/c++config.h
+#        include <bits/c++config.h>
+#        if defined(_GLIBCXX_RELEASE) && _GLIBCXX_RELEASE >= 12
+// libstdc++ does not fully qualify its use of `__is_constant_evaluated`
+// It was introduced in the assert handling in 5e8a30d
+// libstdc++ : Redefine __glibcxx_assert to work in C++ 23 constexpr
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+using ::std::__is_constant_evaluated;
+_LIBCUDACXX_END_NAMESPACE_STD
+#        endif // _GLIBCXX_RELEASE >= 12
+#        define _CCCL_ASSERT_IMPL(expression, message) __glibcxx_assert(expression)
+#      endif // libstdc++
+#    endif // !_CCCL_CUDA_COMPILER_NVCC
+#  else // ^^^ CCCL_ENABLE_DEVICE_ASSERTIONS ^^^ / vvv !CCCL_ENABLE_DEVICE_ASSERTIONS vvv
+#    define _CCCL_ASSERT_IMPL
+#  endif // !CCCL_ENABLE_DEVICE_ASSERTIONS
+#endif // __CUDA_ARCH__
+
+//! Use the different standard library implementations to implement host side asserts
+#ifndef __CUDA_ARCH__
+#  ifdef CCCL_ENABLE_HOST_ASSERTIONS
+#    if __has_include(<yvals_core.h>) // MSVC uses _STL_VERIFY from <yvals_core.h>
+#      include <yvals_core.h>
+#      define _CCCL_ASSERT_IMPL(expression, message) _STL_VERIFY(expression, message)
+#    elif __has_include(<__assert>) // libc++ uses _LIBCPP_ASSERT from <__assert>
+#      include <__assert>
+#      define _CCCL_ASSERT_IMPL(expression, message) _LIBCPP_ASSERT(expression, message)
+#    else // libstdc++ uses __glibcxx_assert from <bits/c++config.h
+#      include <bits/c++config.h>
```

Comment:
If none of `<yvals_core.h>`, `<__assert>`, or `<bits/c++config.h>` are available, then compilation will fail with an error about `<bits/c++config.h>` not being found.

If you want compilation to fail if CCCL is used in an unsupported or unexpected environment, then it should fail with an `#error` that explains why.

If CCCL assertions should gracefully handle unexpected environment, presumably by disabling assertions, then there should be a check for `<bits/c++config.h>` and `_CCCL_ASSERT_IMPL` should be a no-op if that check fails.

Reply:
Thanks a lot, I have added an error message for that